### PR TITLE
vdk-impala: Error handling and recovery support

### PR DESCRIPTION
**Pull request by gageorgiev**
Database connection plugins should be able to handle and recover
from errors in the best way possible.
This change introduces error handling and error recovery to the
vdk-impala plugin. After this change, in the case of an error,
vdk-impala will analyze the error and either retry the query if
possible, execute any necessary additional operation, or print
a descriptive message.
Addresses #472 and #588 

Testing done: introduced unit tests, tested error handling
functionality locally

Signed-off-by: Gabriel Georgiev <gageorgiev@vmware.com>

### CODE REVIEW

**Comment by ivakoleva** [comment]
Did checkout and run tests locally. 
```
(versatile-data-kit) ikoleva@ikoleva-a02 vdk-impala % pytest --junitxml=tests.xml --cov . --cov-report term-missing --cov-report xml:coverage.xml
Test session starts
collecting ... 
 tests/impala_handle_metadata_exception_with_invalidate_and_retry_test.py ✓✓                                                                                                                            7% ▊         
 tests/impala_handle_hdfs_error_test.py ✓✓✓✓✓✓✓✓                                                                                                                                                       37% ███▊      
 tests/impala_handle_should_not_retry_error_test.py ✓                                                                                                                                                  41% ████▏     
 tests/impala_handle_network_error_test.py ✓✓✓✓✓✓                                                                                                                                                      63% ██████▍   
 tests/impala_error_handler_test.py ✓✓✓✓✓✓✓✓✓                                                                                                                                                          96% █████████▋
 tests/test_impala_connection.py ✓                                                                                                                                                                    100% ██████████
================================================================================================= warnings summary ==================================================================================================
../../vdk-core/src/vdk/internal/builtin_plugins/run/file_based_step.py:3
  /Users/ikoleva/IdeaProjects/versatile-data-kit/projects/vdk-core/src/vdk/internal/builtin_plugins/run/file_based_step.py:3: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
    import imp

-- Docs: https://docs.pytest.org/en/stable/warnings.html
--------------------------------------------------- generated xml file: /Users/ikoleva/IdeaProjects/versatile-data-kit/projects/vdk-plugins/vdk-impala/tests.xml ----------------------------------------------------

---------- coverage: platform darwin, python 3.7.8-final-0 -----------
Name                                                                       Stmts   Miss  Cover   Missing
--------------------------------------------------------------------------------------------------------
setup.py                                                                       4      4     0%   3-10
src/vdk/plugin/impala/impala_connection.py                                    27      0   100%
src/vdk/plugin/impala/impala_error_handler.py                                136     16    88%   20, 144, 157, 222-231, 245, 254, 272, 328, 358-364, 367
src/vdk/plugin/impala/impala_plugin.py                                        47      4    91%   153-162
tests/conftest.py                                                             23      2    91%   22-23
tests/impala_error_handler_test.py                                            83      1    99%   248
tests/impala_handle_hdfs_error_test.py                                        80      1    99%   232
tests/impala_handle_metadata_exception_with_invalidate_and_retry_test.py      28      0   100%
tests/impala_handle_network_error_test.py                                     55      1    98%   150
tests/impala_handle_should_not_retry_error_test.py                            17      0   100%
tests/test_impala_connection.py                                               25      0   100%
--------------------------------------------------------------------------------------------------------
TOTAL                                                                        525     29    94%
Coverage XML written to file coverage.xml


Results (600.50s):
      27 passed
```

**Comment by ivakoleva** [comment]
`tests/test_impala_connection.py` seems missing two scenarios covering Impala connectivity error handling. These are
`test_error_handler_exception_handled` and `test_error_handler_exception_not_handled`.

**Comment by gageorgiev** [comment]
> `tests/test_impala_connection.py` seems missing two scenarios covering Impala connectivity error handling. These are `test_error_handler_exception_handled` and `test_error_handler_exception_not_handled`.

Noted, I'd missed that test file when porting them over.

File: projects/vdk-plugins/vdk-impala/tests/impala_error_handler_test.py [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+        self._query = "select 1"
+        self.error_handler = ImpalaErrorHandler(log=logging.getLogger())
+
+    # do not remove patched_time_sleep, it's used by @patch('time.sleep', return_value=None)
+    def test_default_retry(self, patched_time_sleep):
+        test_exception = OperationalError(
+            "Network error"
+        )  # we know for sure this is handled
+        (
+            mock_native_cursor,
```
Comment (ivakoleva) [inline projects/vdk-plugins/vdk-impala/tests/impala_error_handler_test.py:27]: A formatting suggestion, would consider for readability purposes
```suggestion
        mock_native_cursor, _, _, \
        mock_recovery_cursor, _, _, _ = populate_mock_managed_cursor(
```
to focus left-to-right readers on interface surface utilised. 
I would rather simply
```suggestion
        mock_native_cursor, _, _, mock_recovery_cursor, _, _, _ = populate_mock_managed_cursor(
```

**Review by ivakoleva — APPROVED** [review]

**Review by ivakoleva — CHANGES_REQUESTED** [review]
I have requested a buddy for an additional review

**Review by tozka — APPROVED** [review]

**Pull request merged**